docs/system-documentation/docker/configuration.md
# Basic configuration

## Environment Variables

The following environment variables are defined within the environment proposed in the [openbis-app.env and docker-compose templates provided by us](https://sissource.ethz.ch/sispub/openbis-continuous-integration/-/tree/master/hub/openbis-app/compose).

| Variable | Default value | Description |
| -------- | ------------- | ----------- |
|`OPENBIS_ADMIN_PASS`|`123456789`|Administrator password to openBIS instance.|
|`OPENBIS_DATA`|`/data/openbis`|Directory for openBIS persistent data.|
|`OPENBIS_DB_ADMIN_PASS`|`mysecretpassword`|PostgreSQL superuser password. |
|`OPENBIS_DB_ADMIN_USER`|`postgres`|PostgreSQL superuser name.|
|`OPENBIS_DB_APP_PASS`|`mysecretpassword`|Password for application user connecting to the database.|
|`OPENBIS_DB_APP_USER`|`openbis`|Username for application user connecting to the database.|
|`OPENBIS_DB_HOST`|`openbis-db`|Name of container running PostgreSQL database.|
|`OPENBIS_ETC`|`/etc/openbis`|Directory for openBIS configuration files.|
|`OPENBIS_HOME`|`/home/openbis`|Directory for openBIS installation binaries.|
|`OPENBIS_LOG`|`/var/log/openbis`|Directory for openBIS log files.|
|`OPENBIS_FQDN`|`openbis.domain`|Full qualified domain name of openBIS service.|

## Configuration through OS environment

Starting from openBIS 20.10.10, it is now possible to pass all [Java properties](../configuration/configuration-properties-by-module.md) (documented [here](../configuration/index.rst)) as OS environment variables. To do so, either pass those as a list below the `environment:` service element, or within one or several file(s) enlisted below the `env_file:` service element in the docker-compose file. For more information, refer to the official docs on [docs.docker.com](https://docs.docker.com/compose/how-tos/environment-variables/set-environment-variables/).

Since openBIS 20.10.10, hard-coded Java properties of core plugins shipped with the installation can also be modified. The naming of the variables follows the schema `<core-plugin>.<server-type>.<plugin-type>.<plugin-name>.<plugin-property-name>`.

## Configuration through adjusting service.properties files

This is the taditional procedure. It requires that the configuration directories containing the properties file(s) to be changed (created) are mounted as a docker volume on the host. Following our docker-compose templates provided [here](https://sissource.ethz.ch/sispub/openbis-continuous-integration/-/tree/master/hub/openbis-app/compose), the volume to be mounted is the one named `openbis-app-etc`.

To adjust a property specified in, e.g., the AS service.properties, modify the file in-place (assuming docker's data-root path matches the default, `/var/lib/docker`): `edit /var/lib/docker/volumes/openbis-app-config/_data/as/service.properties`.

### Custom configuration files

openBIS offers the ability to pass in configuration files like, e.g., capabilities files. Those can be deployed in any directory mounted as a volume in the openBIS docker container. It needs to be ensured that the associated AS and DSS properties are pointing to the correct file paths.

```{note}
It is necessary to store any data files that needs to be preserved inside the `openbis-app-data` volume, or inside some other docker volume. Likewise, any configuration files should be stored within the `openbis-app-etc` volume. Failure to do so yields undesired behavior in the sense that changes made to the files within the container are being lost when the container goes down.
```

#### Examples

##### Suppy a json file for storing personal access tokens

1. Enable the feature

Add the following line to the AS service.properties:
```
personal-access-tokens-enabled = true
```

2. Ensure the AS property `personal-access-tokens-file-path` is pointing to the correct path to where the json file is located

Assuming the DSS root-dir points to the default dir, `/data/openbis`, then the personal-access-tokens.json could be stored in this directory:
```
personal-access-tokens-file-path = /data/openbis/personal-access-tokens.json
```

3. Create a PAT and monitor the contents of the json file

Described [here](../../software-developer-documentation/apis/personal-access-tokens.md#typical-application-workflow).

##### Suppy a custom capabilities file

To create or modify the capability role map, create a file named `capabilities` within the `as/` sub-directory of the docker volume `openbis-app-etc`.

### Core Plugins

It is possible to make adjustments to core-plugins shipped with the openBIS installer. To do so, just start up openBIS at least once. This will copy the contents of the core-plugins directory to a sub-directory `core-plugins` which is stored within the docker volume `openbis-app-etc`. Any customizations made here will persist restarts of the application, as well as upgrades of the openbis-docker image.

```{warning}
Be careful when making changes to code of core-plugins since they might break when new releases are published. Please consider reading the [Change Log published with each release](https://unlimited.ethz.ch/display/openbis/Production+Releases).

If the application fails to start after changes to the core-plugins have been made, you can always revert to the original state of the core-plugins by removing the `core-plugins` folder within the `openbis-app-etc` volume.
```

Besides adjustments to existing plugins, it is also possible to [create new plugins from scratch](../../software-developer-documentation/server-side-extensions/core-plugins.md).

#### Examples

##### Customize the InstanceProfile.js

This file is part of the `eln-lims` core-plugin. It is located here:
`<openbis-app-etc>/core-plugins/eln-lims/src/as/webapps/eln-lims/html/etc/InstanceProfile.js`

1. Make any changes to this file

E.g., change `this.showSemanticAnnotations = false;` to `this.showSemanticAnnotations = true;`.

2. Restart the container

Most changes made to the configuration of the openBIS application require a restart in order to be applied. Assuming the container running the openBIS application is named `openbis-app`, this is achieved as follows:
```
docker restart openbis-app
```


docs/system-documentation/configuration/afs-configuration.md
AFS Server Configuration
========================

### Folder Structure

| File  | Description                                                                                               |
|-------|-----------------------------------------------------------------------------------------------------------|
| `bin` | Contains one file, `afs_server.sh` that can be used for start, stop, restart or get status of the server. |
| `lib` | Compiled server libraries                                                                                 |
| `log` | Contains log, by default one log file `afs_server.log`.                                                   |
| `etc` | config files                                                                                              |

### etc config files

| File                 | Description                                                                     |
|----------------------|---------------------------------------------------------------------------------|
| `afs_server.conf`    | Java configuration used to start the serve, particularly the amount of RAM used |
| `logging.properties` | Logging configuration, the possible switches are documented on the file         |
| `service.properties` | General configuration                                                           |

#### afs_server.conf

By default, the RAM usage is 512MB, makes assumptions for a small instance using only the web UI.

A good number is: httpMaxContentLength * 3 * Web UI uploads + ttpMaxContentLength * 12 * openBIS Drive uploads

For 10 concurrent users doing uploads using the defaults that means: 15728640 * 3 * 10 = 470~ MB If only using the web UI

#### logging.properties

```properties
#############################################################
# JUL Log Levels (must use these exact names) and Log4j mapping
#
# OFF     = no logging                      | Log4j: OFF
# SEVERE  = serious failures, system errors | Log4j: ERROR
# WARNING = potential problems, recoverable | Log4j: WARN
# INFO    = general informational messages   | Log4j: INFO
# CONFIG  = static configuration messages    | Log4j: DEBUG
# FINE    = general debugging/tracing        | Log4j: DEBUG
# FINER   = more detailed tracing            | Log4j: TRACE
# FINEST  = most detailed, very verbose      | Log4j: TRACE
# ALL     = enable all logging               | Log4j: ALL
#############################################################

# Global logging level
.global.level=INFO

# List of handlers to configure
.global.handlerAliases = myFileHandler, resolvedPropertiesHandler

# Custom File Handler using your custom class
myFileHandler.class = ch.ethz.sis.shared.log.standard.handlers.DailyRollingFileHandler
myFileHandler.logFileName = log/afs_server.log
myFileHandler.maxLogFileSize = 10485760
myFileHandler.append = true
myFileHandler.level = INFO
myFileHandler.messagePattern = %d{yyy-MM-dd HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n

# Resolved Properties Handler
resolvedPropertiesHandler.class = ch.ethz.sis.shared.log.standard.handlers.SingleFileHandler
resolvedPropertiesHandler.level = INFO
resolvedPropertiesHandler.append = false
resolvedPropertiesHandler.maxLogFileSize = 10048576
resolvedPropertiesHandler.filter = OPERATION.ExposablePropertyPlaceholderConfigurer
resolvedPropertiesHandler.logFileName = log/startup_properties.log
resolvedPropertiesHandler.messagePattern = %d %-5p [%t] %c - %m%n
```

#### service.properties (Mandatory)

The service.properties contains around two dozen properties that SHOULD NOT be modified on standard production environments.

Here are ONLY discussed the ones that SHOULD BE manually configured.

| Property                                   | Description                                                                                                                                                                            |
|--------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `writeAheadLogRoot`                        | Defaults to `$DSS_ROOT_DIR/transactions`, this is the write ahead log of the transactional filesystem, SHOULD be on the same filesystem as the final storage for optional performance. |
| `storageRoot`                              | Same used by the DSS. Defaults to `$DSS_ROOT_DIR/store`                                                                                                                                |
| `storageUuid`                              | Same used by the DSS.                                                                                                                                                                  |
| `poolSize`                                 | Defaults to 50, a good number is: Concurrent Uploads/Downloads Web UI + 4 * Concurrent Uploads/Downloads openBIS Drive                                                                 |
| `apiServerInteractiveSessionKey`           | Secret password, to be shared only to those allowed to call BEGIN and PREPARE and COMMIT manually. Used for 2 Phase Transactions                                                       |
| `apiServerTransactionManagerKey`           | Secret password, to be shared only with the transaction manager to run recovery workflows.                                                                                             |
| `openBISUrl`                               | Ideally points directly to the AS in localhost http://localhost:8080                                                                                                                   |
| `openBISTimeout`                           | Defaults to 30000 millis, 30 sec                                                                                                                                                       |
| `openBISUser`                              | Used by the integration plugin, should be an INSTANCE_ADMIN                                                                                                                            |
| `openBISPassword`                          | Password for the openBIS user                                                                                                                                                          |
| `openBISLastSeenDeletionFile`              | Defaults to ./last-seen-deletion                                                                                                                                                       |
| `openBISLastSeenDeletionBatchSize`         | Defaults to 1000                                                                                                                                                                       |
| `openBISLastSeenDeletionIntervalInSeconds` | Defaults to 900 seconds, 15 minutes                                                                                                                                                    |

#### service.properties (Optional)

Additionally, these are MANDATORY for ```Archiving``` but RECOMMENDED for EVERYONE, provides a cache for the immutable dataset metadata.

Values are already filled in the template, just uncomment them.

| Property                          | Description                                                                                 |
|-----------------------------------|---------------------------------------------------------------------------------------------|
| `maintenance-plugins`             | Value to be used: pathInfoFeedingTask                                                       |
| `pathInfoFeedingTask.class`       | Value to be used: ch.systemsx.cisd.etlserver.path.PathInfoDatabaseFeedingTask               |
| `pathInfoFeedingTask.interval`    | Value to be used: 3600                                                                      |
| `pathInfoDB.name`                 | Value to be used: pathinfo                                                                  |
| `pathInfoDB.kind`                 | Value to be used: prod                                                                      |
| `pathInfoDB.engine`               | Value to be used: postgresql                                                                |
| `pathInfoDB.version-holder-class` | Value to be used: ch.systemsx.cisd.openbis.dss.generic.shared.PathInfoDatabaseVersionHolder |
| `pathInfoDB.script-folder`        | Value to be used: afs-server/sql/pathinfo                                                   |

Additionally, these are MANDATORY for setups with multiple share ids, this is for backwards compatibility with the way DSS shuffles data.

| Property                       | Description                                                                      |
|--------------------------------|----------------------------------------------------------------------------------|
| `maintenance-plugins`          | Value to be used: pathInfoFeedingTask, shufflingTask                             |
| `shuffling-task.class`         | Value to be used: ch.systemsx.cisd.etlserver.plugins.SegmentedStoreShufflingTask |
| `shuffling-task.interval`      | Value to be used: 3600                                                           |
| `shuffling.class`              | Value to be used: ch.systemsx.cisd.etlserver.plugins.SimpleShuffling             |
| `shuffling.share-finder.class` | Value to be used: ch.systemsx.cisd.openbis.dss.generic.shared.SimpleShareFinder  |
| `shuffling.verify-checksum `   | Value to be used: true                                                           |

RELEASES.md
# Mockup release 7.0

## Installer

Installing Prerequisites:
- JDK 21
- Postgres 17

[openbis-installer-7.0.tar.gz]()

## Release notes

**Pre-Upgrade Checklist (Example from 20.10.12)**

We’ve resolved an issue that, in rare cases, could allow cycles to be created. To ensure a smooth upgrade, we recommend running the tool below beforehand.
- If the tool doesn’t report anything, you’re all set.
- If it does find something, please address it before upgrading.

This step is important because openBIS 20.10.12 may not start properly if cycles remain in the database after the upgrade.

Run the tool from the command line:
```
java -jar app-openbis-cycle-finder.jar
```

Download it here: [app-openbis-cycle-finder.tar.xz]()

## Source Code
- [Source code (tar.gz)]()
- [Source code (tar)]()

## Development Libraries
- [openbis-java-api-7.0-dependencies-included.tar.gz]()
- [openbis-java-api-7.0.tar.gz]()
- [openbis-javascript-api-7.0-esm.tar.gz]()
- [openbis-python3-api-7.0.tar.gz]()

## Clients
- [openbis-command-line-tool-7.0.tar.gz]()
- [openbis-drive-7.0.tar.gz]()

## Useful Links
- [Documentation]()
- [Change Log]()

server-original-data-store/source/java/ch/systemsx/cisd/etlserver/Parameters.java
/*
 * Copyright ETH 2009 - 2023 Zürich, Scientific IT Services
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ch.systemsx.cisd.etlserver;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Properties;

import org.apache.commons.lang3.time.DateUtils;
import ch.ethz.sis.shared.log.classic.impl.Logger;

import ch.systemsx.cisd.args4j.CmdLineException;
import ch.systemsx.cisd.args4j.CmdLineParser;
import ch.systemsx.cisd.args4j.ExampleMode;
import ch.systemsx.cisd.args4j.Option;
import ch.systemsx.cisd.common.exceptions.ConfigurationFailureException;
import ch.systemsx.cisd.common.exceptions.HighLevelException;
import ch.ethz.sis.shared.log.classic.core.LogCategory;
import ch.ethz.sis.shared.log.classic.impl.LogFactory;
import ch.systemsx.cisd.common.mail.JavaMailProperties;
import ch.systemsx.cisd.common.maintenance.MaintenanceTaskParameters;
import ch.systemsx.cisd.common.maintenance.MaintenanceTaskUtils;
import ch.systemsx.cisd.common.properties.ExtendedProperties;
import ch.systemsx.cisd.common.properties.PropertyParametersUtil;
import ch.systemsx.cisd.common.properties.PropertyParametersUtil.SectionProperties;
import ch.systemsx.cisd.common.properties.PropertyUtils;
import ch.systemsx.cisd.common.time.TimingParameters;
import ch.systemsx.cisd.common.utilities.IExitHandler;
import ch.systemsx.cisd.common.utilities.SystemExit;
import ch.systemsx.cisd.common.utilities.VoidExitHandler;
import ch.systemsx.cisd.openbis.dss.BuildAndEnvironmentInfo;
import ch.systemsx.cisd.openbis.dss.generic.shared.Constants;
import ch.systemsx.cisd.openbis.dss.generic.shared.utils.DssPropertyParametersUtil;

/**
 * The class to process the command line parameters and service properties.
 * 
 * @author Bernd Rinn
 */
public class Parameters
{
    private static final String CHECK_INTERVAL_NAME = "check-interval";

    private static final String QUIET_PERIOD_NAME = "quiet-period";

    private static final Logger operationLog =
            LogFactory.getLogger(LogCategory.OPERATION, Parameters.class);

    private static final Logger notificationLog =
            LogFactory.getLogger(LogCategory.NOTIFY, Parameters.class);

    @Option(name = "s", longName = "server-url", metaVar = "URL", usage = "URL of the server")
    private String serverURL;

    /**
     * The interval to wait between to checks for activity (in milliseconds).
     */
    @Option(name = "c", longName = CHECK_INTERVAL_NAME, usage = "The interval to wait between two checks (in seconds) "
            + "[default: 120]")
    private long checkIntervalSeconds;

    /**
     * Valid only when {@link ThreadParameters#INCOMING_DATA_COMPLETENESS_CONDITION} is false.<br>
     * The period to wait before a file or directory is considered "quiet" (in milliseconds). This setting is used when deciding whether a file or
     * directory is ready to be processed.
     */
    @Option(name = "q", longName = QUIET_PERIOD_NAME, usage = "The period of no write access that needs to pass before an incoming data item is "
            + "considered complete and ready to be processed (in seconds) [default: 300]. "
            + "Valid only when auto-detection method is used to determine if an incoming data are ready to be processed.")
    private long quietPeriodMillis;

    /**
     * The time-out for clean up work in the shutdown sequence (in seconds).
     * <p>
     * Note that that the maximal time for the shutdown sequence to complete can be as large as twice this time.
     */
    @Option(name = "t", longName = "shutdown-timeout", usage = "The time-out for clean up work "
            + "in the shutdown sequence (in seconds) [default: 30]")
    private long shutdownTimeOutSeconds;

    /**
     * The username to access the LIMS server with.
     */
    @Option(name = "u", longName = "username", usage = "User login name")
    private String username;

    /**
     * The password to access the LIMS server with.
     */
    @Option(name = "p", longName = "password", usage = "User login password")
    private String password;

    /**
     * The command line parser.
     */
    private final CmdLineParser parser = new CmdLineParser(this);

    private final TimingParameters timingParameters;

    private final Properties serviceProperties;

    private final ThreadParameters[] threads;

    private MaintenanceTaskParameters[] maintenancePlugins;

    @Option(longName = "help", skipForExample = true, usage = "Prints out a description of the options.")
    void printHelp(final boolean exit)
    {
        parser.printHelp("etlserver", "<required options> [option [...]]", "", ExampleMode.ALL);
        if (exit)
        {
            System.exit(0);
        }
    }

    @Option(longName = "version", skipForExample = true, usage = "Prints out the version information.")
    void printVersion(final boolean exit)
    {
        System.err
                .println("etlserver version " + BuildAndEnvironmentInfo.INSTANCE.getFullVersion());
        if (exit)
        {
            System.exit(0);
        }
    }

    @Option(longName = "test-notify", skipForExample = true, usage = "Tests the notify log (i.e. that an email is "
            + "sent out).")
    void sendTestNotification(final boolean exit)
    {
        notificationLog
                .error("This is a test notification given due to specifying the --test-notify option.");
        if (exit)
        {
            System.exit(0);
        }
    }

    /**
     * This factory method is used by the DSS API to get information about the threads that have been configured. Since the validity of the parameters
     * has already been verified by the ETLServer startup, it does not need to be verified again.
     */
    public static Parameters createParametersForApiUse()
    {
        return new Parameters(new String[0], VoidExitHandler.INSTANCE, false);
    }

    Parameters(final String[] args)
    {
        this(args, SystemExit.SYSTEM_EXIT);
    }

    Parameters(final String[] args, final IExitHandler systemExitHandler)
    {
        this(args, systemExitHandler, true);
    }

    Parameters(final String[] args, final IExitHandler systemExitHandler, boolean checkParameters)
    {
        try
        {
            this.serviceProperties = DssPropertyParametersUtil.loadServiceProperties();
            PropertyUtils.trimProperties(serviceProperties);
            this.threads = createThreadParameters(serviceProperties);
            this.timingParameters = TimingParameters.create(serviceProperties);
            this.maintenancePlugins =
                    MaintenanceTaskUtils.createMaintenancePlugins(serviceProperties);

            initCommandLineParametersFromProperties();

            parser.parseArgument(args);
            if (checkParameters)
            {
                ensureParametersCorrect();
            }
        } catch (final Exception ex)
        {
            operationLog.error(ex.getMessage(), ex);

            outputException(ex);

            if (SystemExit.SYSTEM_EXIT.equals(systemExitHandler))
            {
                operationLog.error("DSS is shutting down.");
            }

            systemExitHandler.exit(1);
            // Only reached in unit tests.
            throw new AssertionError(ex.getMessage());
        }
    }

    public MaintenanceTaskParameters[] getMaintenancePlugins()
    {
        return maintenancePlugins;
    }

    private void ensureParametersCorrect()
    {
        for (final ThreadParameters thread : threads)
        {
            thread.check();
        }

        if (serverURL == null)
        {
            throw new ConfigurationFailureException("No 'server-url' defined.");
        }
    }

    private void initCommandLineParametersFromProperties()
    {
        serverURL = serviceProperties.getProperty("server-url");
        username = serviceProperties.getProperty("username");
        password = serviceProperties.getProperty("password");
        checkIntervalSeconds =
                Long.parseLong(serviceProperties.getProperty(CHECK_INTERVAL_NAME, "120"));
        quietPeriodMillis = Long.parseLong(serviceProperties.getProperty(QUIET_PERIOD_NAME, "300"));
        shutdownTimeOutSeconds =
                Long.parseLong(serviceProperties.getProperty("shutdown-timeout", "30"));
    }

    private static ThreadParameters[] createThreadParameters(final Properties serviceProperties)
    {
        SectionProperties[] sectionsProperties =
                PropertyParametersUtil.extractSectionProperties(serviceProperties,
                        Constants.INPUT_THREAD_NAMES, true);
        if (sectionsProperties.length == 0)
        {
            return new ThreadParameters[0];
        } else
        {
            return asThreadParameters(sectionsProperties);
        }
    }

    private static ThreadParameters[] asThreadParameters(SectionProperties[] sectionProperties)
    {
        List<ThreadParameters> threadParameters = new ArrayList<>();
        for (SectionProperties section : sectionProperties)
        {
            String incomingDirs = section.getProperties().getProperty("incoming-dirs");
            if (incomingDirs == null)
            {
                operationLog.info("Create parameters for thread '" + section.getKey() + "'.");
                threadParameters.add(new ThreadParameters(section.getProperties(), section.getKey()));
            } else
            {
                List<String> dirs = new ArrayList<>(new LinkedHashSet<>(Arrays.asList(incomingDirs.split(","))));
                for (int i = 0; i < dirs.size(); i++)
                {
                    Properties p = (Properties) section.getProperties().clone();
                    String incomingDir = dirs.get(i).trim();
                    p.setProperty(ThreadParameters.INCOMING_DIR, incomingDir);
                    String key = section.getKey() + "[" + (i + 1) + "]";
                    operationLog.info("Create parameters for thread '" + key 
                            + "' (incoming-dir: " + incomingDir + ").");
                    threadParameters.add(new ThreadParameters(p, key));
                }
            }
        }
        return threadParameters.toArray(new ThreadParameters[0]);
    }

    public final static Properties createMailProperties(final Properties serviceProperties)
    {
        final Properties properties =
                ExtendedProperties.getSubset(serviceProperties, "mail", false);
        if (properties.getProperty(JavaMailProperties.MAIL_SMTP_HOST) == null)
        {
            properties.setProperty(JavaMailProperties.MAIL_SMTP_HOST, "localhost");
        }
        if (properties.getProperty(JavaMailProperties.MAIL_FROM) == null)
        {
            properties.setProperty(JavaMailProperties.MAIL_FROM, "etlserver@localhost");
        }
        return properties;
    }

    private void outputException(final Exception ex)
    {
        if (ex instanceof HighLevelException || ex instanceof CmdLineException)
        {
            System.err.println(ex.getMessage());
        } else
        {
            System.err.println("An exception occurred.");
            ex.printStackTrace();
        }
        if (ex instanceof CmdLineException)
        {
            printHelp(false);
        }
    }

    /**
     * Returns The interval to wait between to checks for activity (in milliseconds).
     */
    public long getCheckIntervalMillis()
    {
        return checkIntervalSeconds * DateUtils.MILLIS_PER_SECOND;
    }

    public long getQuietPeriodMillis()
    {
        return quietPeriodMillis * DateUtils.MILLIS_PER_SECOND;
    }

    /**
     * Returns the time-out for clean up work in the shutdown sequence (in seconds).
     * <p>
     * Note that that the maximal time for the shutdown sequence to complete can be as large as twice this time.
     */
    public long getShutdownTimeOutMillis()
    {
        return shutdownTimeOutSeconds * DateUtils.MILLIS_PER_SECOND;
    }

    /**
     * Returns the password to access the LIMS server with.
     */
    public String getPassword()
    {
        return password;
    }

    /**
     * Returns the username to access the LIMS server with.
     */
    public String getUsername()
    {
        return username;
    }

    /**
     * Returns all properties.
     */
    public final Properties getProperties()
    {
        return serviceProperties;
    }

    /**
     * Returns the timing parameters for monitored operations.
     */
    public TimingParameters getTimingParameters()
    {
        return timingParameters;
    }

    /**
     * Logs the current parameters to the {@link LogCategory#OPERATION} log.
     */
    public void log()
    {
        if (operationLog.isInfoEnabled())
        {
            for (final ThreadParameters threadParameters : threads)
            {
                threadParameters.log();
            }

            operationLog.info(String.format("Check intervall: %d s.",
                    getCheckIntervalMillis() / 1000));
            operationLog
                    .info(String
                            .format(
                                    "Quiet period (valid if auto-detection method is used to determine incoming data completeness): %d s.",
                                    getQuietPeriodMillis() / 1000));
        }
    }

    /**
     * Returns the URL of the LIMS server.
     */
    public String getServerURL()
    {
        return serverURL;
    }

    public ThreadParameters[] getThreads()
    {
        return threads;
    }

}
